source/server/admin/grouped_stats_request.cc

```diff
@@ -0,0 +1,156 @@
+
+  // the "type" query param is ignored for prometheus stats, so always start from
+  // counters; also, skip the TextReadouts phase unless that stat type is explicitly
+  // requested via query param
+  if (params_.prometheus_text_readouts_) {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::TextReadouts, "Text Readouts"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  } else {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  }
+  phase_index_ = 0;
+}
+
+template <class StatType> Stats::IterateFn<StatType> GroupedStatsRequest::saveMatchingStat() {
+  return [this](const Stats::RefcountPtr<StatType>& stat) -> bool {
+    // check if unused
+    if (params_.used_only_ && !stat->used()) {
+      return true;
+    }
+
+    // check if filtered
+    if (params_.filter_ != nullptr) {
+      if (!std::regex_search(stat->name(), *params_.filter_)) {
+        return true;
+      }
+    } else if (params_.re2_filter_ != nullptr &&
+               !re2::RE2::PartialMatch(stat->name(), *params_.re2_filter_)) {
+      return true;
+    }
+
+    // capture stat by either adding to a pre-existing variant or by creating a new variant
+    std::string tag_extracted_name = global_symbol_table_.toString(stat->tagExtractedStatName());
+    auto [iterator, inserted] = stat_map_.try_emplace(
+        tag_extracted_name, std::vector<Stats::RefcountPtr<StatType>>({stat}));
+    if (!inserted) {
+      absl::get<std::vector<Stats::RefcountPtr<StatType>>>(iterator->second).emplace_back(stat);
+    }
+    return true;
+  };
+}
+
+Stats::IterateFn<Stats::TextReadout> GroupedStatsRequest::saveMatchingStatForTextReadout() {
+  return saveMatchingStat<Stats::TextReadout>();
+}
+
+Stats::IterateFn<Stats::Gauge> GroupedStatsRequest::saveMatchingStatForGauge() {
+  return saveMatchingStat<Stats::Gauge>();
+}
+
+Stats::IterateFn<Stats::Counter> GroupedStatsRequest::saveMatchingStatForCounter() {
+  return saveMatchingStat<Stats::Counter>();
+}
+
+Stats::IterateFn<Stats::Histogram> GroupedStatsRequest::saveMatchingStatForHistogram() {
+  return saveMatchingStat<Stats::Histogram>();
+}
+
+template <class SharedStatType>
+void GroupedStatsRequest::renderStat(const std::string& name, Buffer::Instance& response,
+                                     const StatOrScopes& variant) {
+  auto prefixed_tag_extracted_name = prefixedTagExtractedName<SharedStatType>(name);
+  if (prefixed_tag_extracted_name.has_value()) {
+    PrometheusStatsRender* const prometheus_render =
+        dynamic_cast<PrometheusStatsRender*>(render_.get());
+    // increment stats count
+    ++phase_stat_count_;
+
+    // sort group
+    std::vector<SharedStatType> group = absl::get<std::vector<SharedStatType>>(variant);
+    std::sort(group.begin(), group.end(),
```

Comment:
perf nit: use SymbolTable::sortByStatName, which is slightly faster because it will hold the symbol table lock over the entire sorting operation, rather than taking the and releasing its lock for each comparison.

Reply:
I tried swapping this with SymbolTable::sortByStatName and I am getting a couple of failures in grouped_stats_requests_test.cc with some lock-related failure (excerpt below)

```
[mutex.cc : 1391] RAW: Acquiring absl::Mutex 0x7fd91a812b28 while holding  0x7fd91a812a18; a cycle in the historical lock ordering graph has been observed
```
I am still digging, though I don't think further reviewing has to block on this.

Reply:
Fixed now!